pkg/profile/artifact_test.go

```diff
@@ -678,4 +681,237 @@ var _ = Describe("Profile", func() {
 			})
 		})
 	})
+	Context("GenerateOutput", func() {
+		It("creates files for all artifacts", func() {
+			pSub = profilesv1.ProfileInstallation{
+				TypeMeta: profileTypeMeta,
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      installationName,
+					Namespace: namespace,
+				},
+				Spec: profilesv1.ProfileInstallationSpec{
+					Source: &profilesv1.Source{
+						URL:    "https://github.com/weaveworks/profiles-examples",
+						Branch: "main",
+						Path:   "weaveworks-nginx",
+					},
+				},
+			}
+			p.SetProfileGetter(func(repoURL, branch, path string, gitClient git.Git) (profilesv1.ProfileDefinition, error) {
+				return profilesv1.ProfileDefinition{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "weaveworks-nginx",
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Profile",
+						APIVersion: "packages.weave.works.io/profilesv1",
+					},
+					Spec: profilesv1.ProfileDefinitionSpec{
+						ProfileDescription: profilesv1.ProfileDescription{
+							Description: "foo",
+						},
+						Artifacts: []profilesv1.Artifact{
+							{
+								Name: "nginx-deployment",
+								Kustomize: &profilesv1.Kustomize{
+									Path: "nginx/deployment",
+								},
+							},
+						},
+					},
+				}, nil
+			})
+			tempDir, err := ioutil.TempDir("", "catalog-install")
+			Expect(err).NotTo(HaveOccurred())
+			maker := profile.NewProfilesArtifactsMaker(profile.MakerConfig{
+				ProfileName:      "generate-test",
+				GitClient:        fakeGitClient,
+				RootDir:          tempDir,
+				GitRepoNamespace: gitRepoNamespace,
+				GitRepoName:      gitRepoName,
+			})
+			artifacts, err := maker.MakeArtifacts(pSub)
+			Expect(err).NotTo(HaveOccurred())
+			fakeGitClient.SparseCloneStub = func(url string, branch string, dir string, p string) error {
+				fullPath := filepath.Join(dir, "weaveworks-nginx", "nginx", "deployment")
+				err := os.MkdirAll(fullPath, 0755)
+				Expect(err).NotTo(HaveOccurred())
+				return ioutil.WriteFile(filepath.Join(fullPath, "deployment.yaml"), []byte("validYaml:"), 0644)
+			}
+			err = maker.GenerateOutput(artifacts, pSub)
```

Comment:
I think this naming is slightly messy. We have an `ArtifactMaker` (whats make in this context? the in memory structs, the files?), which has a `MakeArifacts` (can probably just be `Make` or really in an ideal world just `()` since its implied from struct name). And the `GenerateOutput` which feels misplaced.

I think part of the smell is that its two funcs that should be one, which makes naming the functions and the struct hard. So how about either:
- Moving the file writing functionality outside of this interface, which is probably not ideal since it was moved here in this PR aha :wink: but perhaps theres a slightly different abstraction that games the generating"and writing to files separate
- Merge them to one func, then we have `ArtifactMaker.Make()` which keeps it a bit cleaner. Then we can extract out some underlying parts of the ArtifactMaker into smaller units? wdyt?

Reply:
Struct: `Artifactory` (yes it is an epic idea)
Funcs: `Make` and `Write` (isn't that what it is doing? I don't get where the "generate" comes from)

Reply:
Oh, sorry Claudia, I didn't see your comment. :( I put everything into a single `Make` in the end.

Reply:
No worries 😄  can I ask why you chose to refactor at this level, around the general "make all the things call"? I assumed you would be having the interface at the `makeArtifacts` level and then have each artifact fulfil the interface and get rid of that giant `if else` which is growing by the second.

And I realise I should have asked this in the first PR 🤦‍♀️

Reply:
`makeArtifact` will disappear and profile will be a provider for the Maker. And then profiles will have builders which build the individual artifacts. That's the main idea. :)

Reply:
Perfect!